src/components/TradeWidget/TokenRow.tsx

```diff
@@ -265,42 +211,13 @@ const TokenRow: React.FC<Props> = ({
     setValue(inputId, formatAmountFull(balance.totalExchangeBalance, balance.decimals, false), true)
   }
 
-  const enforcePrecision = useCallback(() => {
-    const newValue = adjustPrecision(inputValue, selectedToken.decimals)
-    if (inputValue !== newValue) {
-      setValue(inputId, newValue, true)
-    }
-  }, [inputValue, selectedToken.decimals, setValue, inputId])
-
-  useEffect(() => {
-    enforcePrecision()
-  }, [enforcePrecision])
-
-  const removeExcessZeros = useCallback(
-    (event: React.SyntheticEvent<HTMLInputElement>): void => {
-      // Q: Why do we need this function instead of relying on `preventInvalidChars` or `enforcePrecision`?
-      // A: Because on those functions we still want the user to be able to input partial values. E.g.:
-      //    0 -> 0. -> 0.1 -> 0.10 -> 0.105
-      //    When losing focus though (`onBlur`), we remove everything that's redundant, such as leading zeros,
-      //    trailing dots and/or zeros
-      // Q: Why not use formatAmount/parseAmount that already take care of this?
-      // A: Too many steps (convert to and from BN) and binds the function to selectedToken.decimals
-
-      const { value } = event.currentTarget
-      const newValue = value.replace(leadingAndTrailingZeros, '').replace(trailingZerosAfterDot, '$1')
-
-      if (value != newValue) {
-        setValue(inputId, newValue, true)
-      }
-    },
-    [inputId, setValue],
-  )
-
-  const onKeyPress = useCallback(
-    (event: React.KeyboardEvent<HTMLInputElement>): void =>
-      event.key === 'Enter' ? removeExcessZeros(event) : preventInvalidChars(event),
-    [removeExcessZeros],
-  )
+  // Form validation
+  const inputRef = !readOnly
```

Comment:
Can `readOnly` change when the component is mounted? Because I don't think it will re-register as react-hook-form does it only once.
Totally ok if `readOnly` is the same throughout the life of the component

Reply:
Readonly don't change, cause it's only used to define the receivingAmount as readOnly. That will remain true forever.

Hoever, as a component, no-one warranties the use you will give to that property, so I agree it can be an issue. It's a bad practice to depend on "how you use the props"

Reply:
Do you see a simple solution?

Reply:
```ts
const inputRef = useMemo(() => !readOnly
   ? register({
        pattern: { value: validInputPattern, message: 'Invalid amount' },
        validate: { positive: validatePositive },
      })
    : register,
// eslint-disable don't-remember-what
[register])
```

Reply:
No, don't include readOnly in deps, the whole point is to not depend on it
And don't copy my eslint comment, lol

Reply:
ahahahah